## Homework - Module-1

### Question:1

In [3]:
import numpy as np
import pandas as pd
pd.__version__

'2.2.2'

Getting the data
For this homework, we'll use the Laptops Price dataset. Download it from here.

You can do it with wget:

wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv
Or just open it with your browser and click "Save as...".

Now read it with Pandas.

### Question:2

In [5]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv')
df

Laptop       Status   Brand  \
0     ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...          New    Asus   
1     Alurin Go Start Intel Celeron N4020/8GB/256GB ...          New  Alurin   
2     ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...          New    Asus   
3     MSI Katana GF66 12UC-082XES Intel Core i7-1270...          New     MSI   
4     HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...          New      HP   
...                                                 ...          ...     ...   
2155  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2156  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2157  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2158  Razer Book 13 Intel Evo Core i7-1165G7/16GB/1T...  Refurbished   Razer   
2159  Razer Book FHD+ Intel Evo Core i7-1165G7/16GB/...  Refurbished   Razer   

           Model                CPU  RAM  Storage Storage type       GPU  \
0     ExpertBook      Intel Core i5    8      512          SSD       NaN   
1             Go      Intel Celeron    8      256          SSD       NaN   
2     ExpertBook      Intel Core i3    8      256          SSD       NaN   
3         Katana      Intel Core i7   16     1000          SSD  RTX 3050   
4            15S      Intel Core i5   16      512          SSD       NaN   
...          ...                ...  ...      ...          ...       ...   
2155       Blade      Intel Core i7   16     1000          SSD  RTX 3060   
2156       Blade      Intel Core i7   16     1000          SSD  RTX 3070   
2157       Blade      Intel Core i7   32     1000          SSD  RTX 3080   
2158        Book  Intel Evo Core i7   16     1000          SSD       NaN   
2159        Book  Intel Evo Core i7   16      256          SSD       NaN   

      Screen Touch  Final Price  
0       15.6    No      1009.00  
1       15.6    No       299.00  
2       15.6    No       789.00  
3       15.6    No      1199.00  
4       15.6    No       669.01  
...      ...   ...          ...  
2155    17.3    No      2699.99  
2156    17.3    No      2899.99  
2157    17.3    No      3399.99  
2158    13.4   Yes      1899.99  
2159    13.4   Yes      1699.99  

[2160 rows x 12 columns]

In [18]:
df.columns
df['Brand'].value_counts()

Brand
Asus                415
HP                  368
Lenovo              366
MSI                 308
Acer                137
Apple               116
Dell                 84
Microsoft            77
Gigabyte             48
Razer                37
Medion               32
LG                   32
Alurin               29
PcCom                24
Samsung              22
Dynabook Toshiba     19
Vant                 11
Primux                8
Deep Gaming           8
Innjoo                6
Thomson               4
Prixton               3
Millenium             2
Denver                1
Jetwing               1
Realme                1
Toshiba               1
Name: count, dtype: int64

In [7]:
df.shape[0]

2160

### Question:3

Q3. Laptop brands
How many laptop brands are presented in the dataset?

12
27
28
2160

In [14]:
df['Brand'].nunique()

27

### Question:4
Q4. Missing values
How many columns in the dataset have missing values?

0
1
2
3

In [13]:
df.isnull().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

3

### Question 5
Q5. Maximum final price
What's the maximum final price of Dell notebooks in the dataset?

869
3691
3849
3936

In [20]:
df[df['Brand'] == 'Dell']['Final Price'].max()

3936.0

### Question 6
Q6. 
Median value of Screen

Find the median value of Screen column in the dataset.

Next, calculate the most frequent value of the same Screen column.

Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.

Now, calculate the median value of Screen once again.

In [21]:
df['Screen'].median()

15.6

In [24]:
screen_mode_value = df['Screen'].mode()
screen_mode_value

0    15.6
Name: Screen, dtype: float64

In [25]:
df['Screen'].fillna(screen_mode_value)

0       15.6
1       15.6
2       15.6
3       15.6
4       15.6
        ... 
2155    17.3
2156    17.3
2157    17.3
2158    13.4
2159    13.4
Name: Screen, Length: 2160, dtype: float64

In [26]:
df['Screen'].fillna(screen_mode_value).median()

15.6

### Question 7

Q7. Sum of weights

Select all the "Innjoo" laptops from the dataset.

Select only columns RAM, Storage, Screen.

Get the underlying NumPy array. Let's call it X.

Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

Compute the inverse of XTX.

Create an array y with values [1100, 1300, 800, 900, 1000, 1100].

Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

What's the sum of all the elements of the result?


In [27]:
innjoo_df = df[df['Brand'] == 'Innjoo']
innjoo_df = innjoo_df[['RAM', 'Storage', 'Screen']]
innjoo_df

RAM  Storage  Screen
1478    8      256    15.6
1479    8      512    15.6
1480    4       64    14.1
1481    6       64    14.1
1482    6      128    14.1
1483    6      128    14.1

In [28]:
X = innjoo_df.values
X

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1],
       [  6. , 128. ,  14.1]])

In [29]:
XTX = X.T.dot(X)

In [31]:
## inverse of XTX
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 2.78025381e-01, -1.51791334e-03, -1.00809855e-01],
       [-1.51791334e-03,  1.58286725e-05,  4.48052175e-04],
       [-1.00809855e-01,  4.48052175e-04,  3.87214888e-02]])

In [33]:
y = np.array([1100, 1300, 800, 900, 1000, 1100])
y

array([1100, 1300,  800,  900, 1000, 1100])

In [34]:
w = XTX_inv.dot(X.T).dot(y)
w

array([45.58076606,  0.42783519, 45.29127938])

In [36]:
w.sum()

91.2998806299557